docs/source/advanced/splitters.rst

```diff
@@ -0,0 +1,168 @@
+.. _splitters:
+
+Splitters
+=========
+
+.. contents::
+    :local:
+    :backlinks: entry
+
+
+By default outlets will be given all produced records at the same time. Splitters are a middleware which allows you to break that list of records into batches. Each batch is then fed into outlets separately, allowing outlets to process an entire batch individually at the same time, instead of processing each record one by one.
+
+
+
+Simple example
+--------------
+
+Following splitter will split the records into batches based on their payload 'name' attribute.
+
+.. code-block:: python
+
+    def splitter_by_name(batches: List[List[Record]]):
+        result = []
+        for batch in batches:
+            split = {}
+            for record in batch:
+                split[record.payload['name']] = record
+
+            result.append(list(split.values()))
+
+    link = Link(..., splitters=splitter_by_name)
+
+.. rst-class:: mb-s
+
+    Which will turn the following list of records:
+
+.. rst-class:: highlight-small
+.. code-block:: python
+
+    [
+        {'name': 'a', 'value': 1},
+        {'name': 'a', 'value': 2},
+        {'name': 'b', 'value': 3},
+        {'name': 'b', 'value': 4}
+    ]
+
+.. rst-class:: mb-s
+
+    into the following list of batches:
+
+.. rst-class:: highlight-small
+.. code-block:: python
+
+    [
+        [
+            {'name': 'a', 'value': 1},
+            {'name': 'a', 'value': 2}
+        ],
+        [
+            {'name': 'b', 'value': 3},
+            {'name': 'b', 'value': 4}
+        ]
+    ]
+
+
+Splitters explained
+-------------------
+
+A splitter is a :any:`callable` function that accepts a list of batches and returns a list of batches.
```

Comment:
How about something like:

```A splitter is a :any:`callable` function that accepts a list of batches (a batch is a list of :any: `Records <records>`) and returns a list of batches.```  to help clarify that a single batch is itself a list of records.

Reply:
So my logic here was that we zoom in further and further into details. This one line alone works badly like you observe, but it is composed with the following two lines in order to explain the terminology:


> A splitter is a `callable` function that accepts a list of batches and returns a list of batches.
> A *list of batches* is a two-dimensional list containing the records divided into sub-lists.
> Each of these sub-lists is called a *batch*.

So the first line is more overviewing, and immediately after we explain what the 'list of batches' is. Once we do that we explain what a 'batch' is.

If we went ahead with your suggestion we'd essentially be repeating ourselves in the next line, right?

Btw - I do agree that adding :any:`Records` would work better than just saying `records` here.

Reply:
You mentioned in the comments for this example that:
 >  the default batch contains all records

Maybe we could bring this into the first line to add a bit of clarification: 
`A splitter is a :any: callable function that accepts records (the default batch) and returns a list of batches. `

Reply:
> A splitter is a :any: callable function that accepts records (the default batch) and returns a list of batches.

I see what you're trying to accomplish here, unfortunately though that's _sortof_ true only for the first splitter provided and not true for all splitters. Even the first splitter doesn't accept records, but also a list of batches that just happens to have one batch inside of it containing all records (this is mentioned elsewhere in the documentation). In generic terms: a splitter accepts a list of batches, not records. 

I'm not saying we couldn't make this wording better though! Please feel free to suggest improvements to this if you can think of any. I'm merely pointing out that this one in particular is inaccurate, and I can't think of anything better than what I put already.

Reply:
You're right on this, and I can't think of any other wording, this seems to be as precise as it can get for now. Lets go with it.

Reply:
Okay cool - well sorry I refused your suggestions here, and I appreciate you bringing these up, but it really is a tricky section to word correctly. We'll need if it needs rephrasing if we end up having to explain this a lot.